server/http/restapi/embedded_form.go

```diff
@@ -0,0 +1,166 @@
+	if err != nil {
+		writeDialogError(w, "Could not decode request.")
+		return
+	}
+
+	if dialogRequest.Submission[embeddedSubmissionPostIDKey] == nil {
+		writeDialogError(w, "No Post ID provided.")
+		return
+	}
+
+	postID, ok := dialogRequest.Submission[embeddedSubmissionPostIDKey].(string)
+	if !ok {
+		writeDialogError(w, "Post ID not recognized.")
+		return
+	}
+
+	delete(dialogRequest.Submission, embeddedSubmissionPostIDKey)
+
+	if dialogRequest.Submission[embeddedSubmissionAppIDKey] == nil {
+		writeDialogError(w, "No App ID provided")
+		return
+	}
+
+	appID, ok := dialogRequest.Submission[embeddedSubmissionAppIDKey].(string)
+	if !ok {
+		writeDialogError(w, "App ID not recognized")
+		return
+	}
+
+	delete(dialogRequest.Submission, embeddedSubmissionAppIDKey)
+
+	c := &apps.Call{
+		Wish: store.NewWish(appID, dialogRequest.URL),
+		Request: &apps.CallRequest{
+			Context: &apps.Context{
+				AppID:        store.AppID(appID),
+				ActingUserID: dialogRequest.UserId,
+				ChannelID:    dialogRequest.ChannelId,
+				TeamID:       dialogRequest.TeamId,
+				UserID:       dialogRequest.UserId,
+				PostID:       postID,
+			},
+			Values: apps.FormValues{
+				Data: dialogRequest.Submission,
+			},
+		},
+	}
+
+	resp, err := a.apps.Client.PostWish(c)
+	if err != nil {
+		writeDialogError(w, "Error contacting the app: "+err.Error())
+		return
+	}
+
+	var dialogResponse model.SubmitDialogResponse
+
+	if resp.Type == apps.ResponseTypeError {
+		if resp.Data["errors"] != nil {
+			dialogResponse.Errors = make(map[string]string)
+			if errors, ok := resp.Data["errors"].(map[string]interface{}); ok {
+
+				for key, value := range errors {
+					if svalue, ok := value.(string); ok {
+						dialogResponse.Errors[key] = svalue
+					}
+				}
+			}
+		}
+		dialogResponse.Error = resp.Error
+	} else {
+		if resp.Data["post"] != nil {
+			if updatePost, parseErr := postFromInterface(resp.Data["post"]); parseErr == nil {
+				updateErr := a.UpdatePost(postID, updatePost)
+				if updateErr != nil {
+					a.mm.Log.Debug("could not update post", "error", updateErr)
+				}
+			} else {
+				a.mm.Log.Debug("could not transform post", "error", parseErr)
+			}
+		}
+	}
+	httputils.WriteJSON(w, dialogResponse)
+}
+
+func writeDialogError(w http.ResponseWriter, msg string) {
+	httputils.WriteJSON(w, model.SubmitDialogResponse{
+		Error: msg,
+	})
+}
+
+func postFromInterface(v interface{}) (*model.Post, error) {
+	b, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+
+	var post model.Post
+	err = json.Unmarshal(b, &post)
+	if err != nil {
+		return nil, err
+	}
+
+	return &post, nil
+}
+
+func (a *api) UpdatePost(postID string, post *model.Post) error {
+	// If the updated post does contain a replacement Props set, we still
+	// need to preserve some original values, as listed in
+	// model.PostActionRetainPropKeys. remove and retain track these.
+	remove := []string{}
+	retain := map[string]interface{}{}
```

Comment:
Can you link to the source of this comment here? https://github.com/mattermost/mattermost-server/blob/20491c2585475c2218f964e0a882c65deac570a5/app/integration_action.go#L57